src/v/pandaproxy/kafka_client_cache.cc

```diff
@@ -0,0 +1,175 @@
+
+namespace pandaproxy {
+
+namespace {
+template<typename List_Type>
+ss::future<> clean_clients_from_list(
+  List_Type& list,
+  mutex& lock,
+  config::binding<std::chrono::milliseconds> keep_alive) {
+    auto has_expired = timestamped_user::clock::now();
+    auto condition = [keep_alive, has_expired](const timestamped_user& item) {
+        return has_expired >= (item.last_used + keep_alive());
+    };
+
+    auto first = list.begin();
+    auto last = list.end();
+    while (first != last) {
+        if (condition(*first)) {
+            vlog(plog.debug, "Remove {} from cache", first->key);
+            // Stop client before erase
+            co_await lock.with([&first] { return first->client->stop(); });
+            list.erase(first++);
+        } else {
+            ++first;
+        }
+    }
+}
+
+template<typename List_Type>
+ss::future<> stop_clients_until_empty(List_Type& list, mutex& lock) {
+    while (!list.empty()) {
+        co_await lock.with([&list] { return list.back().client->stop(); });
+        list.pop_back();
+    }
+}
+} // namespace
+
+kafka_client_cache::kafka_client_cache(
+  YAML::Node const& cfg,
+  config::binding<size_t> max_size,
+  config::binding<std::chrono::milliseconds> keep_alive)
+  : _config{cfg}
+  , _cache_max_size{max_size}
+  , _keep_alive{keep_alive} {}
+
+client_ptr kafka_client_cache::make_client(
+  credential_t user, config::rest_authn_method authn_method) {
+    kafka::client::configuration cfg{
+      to_yaml(_config, config::redact_secrets::no)};
+
+    // Set the principal when the request is using HTTP Basic AuthN
+    if (authn_method == config::rest_authn_method::http_basic) {
+        // Need to specify type or else bad any_cast runtime error
+        cfg.sasl_mechanism.set_value(ss::sstring{"SCRAM-SHA-256"});
+        cfg.scram_username.set_value(std::move(user.name));
+        cfg.scram_password.set_value(std::move(user.pass));
+    }
+
+    return ss::make_lw_shared<kafka::client::client>(
+      to_yaml(cfg, config::redact_secrets::no));
+}
+
+ss::future<client_ptr> kafka_client_cache::fetch_or_insert(
+  credential_t user, config::rest_authn_method authn_method) {
+    auto& inner_list = _cache.get<underlying_list>();
+    auto& inner_hash = _cache.get<underlying_hash>();
+    ss::sstring k{user.name};
+    auto it_hash = inner_hash.find(k);
+
+    client_ptr client;
+
+    // When no client is found ...
+    if (it_hash == inner_hash.end()) {
+        if (_cache.size() >= _cache_max_size()) {
+            if (!_cache.empty()) {
+                auto item = inner_list.back();
+                vlog(plog.debug, "Cache size reached, evicting {}", item.key);
+                inner_list.pop_back();
+                _evicted_items.push_back(item);
+            }
+        }
+
+        vlog(plog.debug, "Make client for user {}", k);
+
+        client = make_client(user, authn_method);
+        inner_list.emplace_front(k, client);
+    } else {
+        co_await _lock.with([this,
```

Comment:
Spoke with @dotnwat about this. With the eviction list, locking is not needed on the datastructure but instead let us consider what happens when many clients are opened/created for the same user. 

Possible approaches:
1. Add a lock to each timestamped user. On a per-user basis, acquire the lock at the start of client creation and release the lock at the end.
2. Add a single lock to the cache. On a per-fiber basis, acquire the lock at the start of client creation and release the lock at the end.

I've been implementing (2) but the code can be cleaned up. (1) May be a simpler design to implement.